File: mycollab-ui/src/main/scala/com/mycollab/vaadin/mvp/PageActionChain.scala
package com.mycollab.vaadin.mvp

import scala.collection.mutable

/**
  * @author MyCollab Ltd
  * @since 5.0.9
  */
class PageActionChain {
  val chains: mutable.Buffer[ScreenData[_]] = mutable.Buffer()

  def this(pageActionArr: Array[ScreenData[_]]) = {
    this()
    chains.appendAll(pageActionArr)
  }

  def this(param: ScreenData[_]) = this(Array[ScreenData[_]](param))

  def this(param1: ScreenData[_], param2: ScreenData[_]) = this(Array[ScreenData[_]](param1, param2))

  def this(param1: ScreenData[_], param2: ScreenData[_], param3: ScreenData[_]) = this(Array[ScreenData[_]](param1, param2, param3))

  def add(pageAction: ScreenData[_]): PageActionChain = {
    chains += pageAction
    this
  }

  def pop: ScreenData[_] = {
    if (chains.size > 0) {
      val pageAction = chains(0)
      chains.remove(0)
      pageAction
    }
    else null
  }

  def peek: ScreenData[_] = chains(0)

  def hasNext: Boolean = chains.length > 0
}

File: mycollab-ui/src/main/java/com/mycollab/vaadin/mvp/PageActionChain.kt
package com.mycollab.vaadin.mvp

/**
 * @author MyCollab Ltd
 * @since 6.0.0
 */
class PageActionChain(vararg pageActionArr: ScreenData<*>) {
    private val chains = mutableListOf<ScreenData<*>>()

    init {
        chains.addAll(pageActionArr)
    }

    fun add(pageAction: ScreenData<*>): PageActionChain {
        chains.add(pageAction)
        return this
    }

    fun pop(): ScreenData<*>? {
        return if (chains.size > 0) {
            val pageAction = chains[0]
            chains.removeAt(0)
            pageAction
        } else null
    }

    fun peek(): ScreenData<*> = chains[0]

    fun hasNext(): Boolean = chains.size > 0
}